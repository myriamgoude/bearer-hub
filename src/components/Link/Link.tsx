import React from 'react'
import { Link as GatsbyLink } from 'gatsby'
import { isBrowser } from '../../services/Browser'

interface ILinkProps {
  to: string
  trackLink?: boolean
  targetBlank?: boolean
  [key: string]: any
}

class Link extends React.Component<ILinkProps> {
  constructor(props: ILinkProps) {
    super(props)
  }

  // Typically the action (event name) is the internal link (e.g. /explore)
  // but this may be overridden and there are exceptions e.g. for some external links
  private getTrackingAction = () => {
    if (this.props.trackingAction) {
      return this.props.trackingAction
    }
    switch (this.props.to) {
      case 'https://docs.bearer.sh': {
        return 'general-documentation'
      }
      case '/integrations': {
        return 'explore-page'
      }
      case '/product': {
        return 'product-page'
      }
      default: {
        return this.props.to.replace(/\//g, '')
      }
    }
  }

  public handleOnClick = (event: any) => {
    // onClick action to track analytics
    if (this.props.trackLink && isBrowser) {
      const currentWindow = window as any
      currentWindow.analytics.track(this.getTrackingAction(), this.props.trackingOptions || {})
    }

    // other onClick actions (if present)
    if (this.props.onClick) {
      this.props.onClick(event)
    }
  }

  render() {
    const internal = /^\/(?!\/)/.test(this.props.to)
    const { to, trackLink, trackingAction, trackingOptions, onClick, targetBlank, ...remainingProps } = this.props

    if (internal) {
      return (
        <GatsbyLink
          to={this.props.to}
          {...(this.props.targetBlank ? { target: '_blank', rel: 'noopener' } : null)}
          {...(this.props.trackLink || this.props.onClick ? { onClick: this.handleOnClick } : null)}
          {...remainingProps}
        >
          {this.props.children}
        </GatsbyLink>
      )
    }
    return (
      <a
        href={this.props.to}
        {...(this.props.targetBlank ? { target: '_blank', rel: 'noopener' } : null)}
        {...(this.props.trackLink || this.props.onClick ? { onClick: this.handleOnClick } : null)}
        {...remainingProps}
      >
        {this.props.children}
      </a>
    )
  }
}

export default Link
