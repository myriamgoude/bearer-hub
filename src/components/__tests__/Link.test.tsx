import React from 'react'
import * as ShallowRenderer from 'react-test-renderer/shallow'
import Link from '../Link/Link'

describe('Header', () => {
  const text = 'my link'
  const renderer = ShallowRenderer.createRenderer()
  it('renders internal links', () => {
    // Note: Link will appear as mockConstructor in snapshot
    // since its mocked in /__mocks__/gatsby due to its dependacy on gatsby internals
    const to = '/foo'
    renderer.render(<Link to={to}>{text}</Link>)
    expect(renderer.getRenderOutput()).toMatchSnapshot()
  })
  it('renders external links', () => {
    const to = 'http://example.com'
    renderer.render(<Link to={to}>{text}</Link>)
    expect(renderer.getRenderOutput()).toMatchSnapshot()
  })
  it('renders tricky non gatsby links', () => {
    const to = '#'
    renderer.render(<Link to={to}>{text}</Link>)
    expect(renderer.getRenderOutput()).toMatchSnapshot()
  })
  it('passes props to subcomponents', () => {
    const handler = () => {}
    const to = '#'
    renderer.render(
      <Link to={to} onClick={handler}>
        {text}
      </Link>
    )
    expect(renderer.getRenderOutput()).toMatchSnapshot()
  })
})
