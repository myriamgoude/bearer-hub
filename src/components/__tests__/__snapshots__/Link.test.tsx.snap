// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link passes props to subcomponents 1`] = `
<mockConstructor
  onClick={[Function]}
  to="/bar"
>
  my link
</mockConstructor>
`;

exports[`Link renders external links as expected (e.g. target blank) 1`] = `
<a
  href="http://example.com"
  rel="noopener"
  target="_blank"
>
  my link
</a>
`;

exports[`Link renders internal links 1`] = `
<mockConstructor
  to="/foo"
>
  my link
</mockConstructor>
`;
